File: Arrays/121. Best Time to Buy and Sell Stock/121. Best Time to Buy and Sell Stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
  let min_price = Infinity;
  let max_profit = 0;

  for (const p of prices) {
    min_price = Math.min(min_price, p);
    max_profit = Math.max(max_profit, p - min_price);
  }

  return max_profit;
};

// Time: O(n)
// Space: O(1)

File: Arrays/121. Best Time to Buy and Sell Stock/121. Best Time to Buy and Sell Stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_price = float("inf")
        max_profit = 0

        for p in prices:
            min_price = min(min_price, p)
            max_profit = max(max_profit, p - min_price)
        
        return max_profit

# Time: O(n)
# Space: O(1)